pkg/image/registry/imagestream/strategy.go

```diff
@@ -229,19 +230,31 @@ func tagReferenceToTagEvent(stream *api.ImageStream, tagRef api.TagReference, ta
 		}, nil
 
 	case "ImageStreamImage":
-		ref, err := api.DockerImageReferenceForStream(stream)
-		if err != nil {
-			return nil, err
+		var event *api.TagEvent
+		// while searching for all the image IDs in stream that start with imageID
+		// save the tag event for proper tagging, code borrowed from api.ResolveImageID
+		resolvedIDs := sets.NewString()
+		for _, history := range stream.Status.Tags {
+			for _, tagging := range history.Items {
+				if d, err := digest.ParseDigest(tagging.Image); err == nil {
+					if strings.HasPrefix(d.Hex(), tagOrID) || strings.HasPrefix(tagging.Image, tagOrID) {
+						event = &tagging
+						resolvedIDs.Insert(tagging.Image)
+					}
+					continue
+				}
+				if strings.HasPrefix(tagging.Image, tagOrID) {
+					event = &tagging
+					resolvedIDs.Insert(tagging.Image)
+				}
+			}
 		}
-
-		resolvedIDs := api.ResolveImageID(stream, tagOrID)
```

Comment:
`ResolveImageID` is only used in 2 places currently: here, and in imagestreamimage/rest.go. We know what it's used for here. In rest.go, we use it to get the image ID (sha256:...), which we then use to get the actual `Image` (which obviously assumes that the image has been imported into etcd).

Let's see if there's a way to modify this method to suit both purposes...